Classes/common/Indexing/CDTIndexManager.m

```diff
@@ -823,6 +759,132 @@ - (BOOL)isValidIndexName:(NSString *)indexName error:(NSError *__autoreleasing *
+{
+    NSString *dir = [datastore extensionDataFolder:kCDTExtensionName];
+    [[NSFileManager defaultManager] createDirectoryAtPath:dir
+                              withIntermediateDirectories:TRUE
+                                               attributes:nil
+                                                    error:nil];
+    NSString *filename = [NSString pathWithComponents:@[ dir, @"indexes.sqlite" ]];
+
+    FMDatabaseQueue *database = nil;
+    NSError *thisError = nil;
+    BOOL success = YES;
+
+    if (success) {
+        database = [[FMDatabaseQueue alloc] initWithPath:filename];
+
+        success = (database != nil);
+        if (!success) {
+            NSDictionary *userInfo = @{
+                NSLocalizedDescriptionKey :
+                    NSLocalizedString(@"Problem opening or creating database.", nil)
+            };
+            thisError = [NSError errorWithDomain:CDTIndexManagerErrorDomain
+                                            code:CDTIndexErrorSqlError
+                                        userInfo:userInfo];
+        }
+    }
+
+    if (success) {
+        id<CDTEncryptionKeyProvider> provider = [datastore copyEncryptionKeyProvider];
+
+        success = [CDTIndexManager configureDatabase:database withEncryptionKeyProvider:provider];
+
+        if (!success) {
+            NSDictionary *userInfo = @{
+                NSLocalizedDescriptionKey :
+                    NSLocalizedString(@"Problem configuring database with encryption key", nil)
+            };
+            thisError = [NSError errorWithDomain:CDTIndexManagerErrorDomain
+                                            code:CDTIndexErrorEncryptionKeyError
+                                        userInfo:userInfo];
+        }
+    }
+
+    if (success) {
+        success = [CDTIndexManager updateSchema:VERSION inDatabase:database];
+
+        if (!success) {
+            NSDictionary *userInfo = @{
+                NSLocalizedDescriptionKey :
+                    NSLocalizedString(@"Problem updating database schema.", nil)
+            };
+            thisError = [NSError errorWithDomain:CDTIndexManagerErrorDomain
+                                            code:CDTIndexErrorSqlError
+                                        userInfo:userInfo];
+        }
+    }
+
+    if (!success) {
+        database = nil;
+
+        if (error) {
+            *error = thisError;
+        }
+    }
+
+    return database;
+}
+
++ (BOOL)configureDatabase:(FMDatabaseQueue *)database
+withEncryptionKeyProvider:(id<CDTEncryptionKeyProvider>)provider
+{
+    __block BOOL success = YES;
+
+    NSString *key = [provider encryptionKey];
+    if (key) {
+        [database inDatabase:^(FMDatabase *db) {
+          success = [db setKey:key];
+        }];
+
+        if (!success) {
+            CDTLogError(CDTINDEX_LOG_CONTEXT, @"Key to cipher database not set");
+        }
+    }
+
+    return success;
+}
```

Comment:
This has far fewer checks than the version inside TD_Database. E.g.:

- No check for whether the database is encrypted if there is no key provided.
- You don't do the query to `sqlite_master` to make sure the key actually works.

It might be worth pulling out a utility class which does all these checks so both TD_Database and CDTIndexManager (and the Query index manager) can use it.